Implemented string upper/lower case assertions

weitzhandler opened the pull request:
Adds the following `Should`s to `string`, to assert its casing:

- `BeUpper`
- `NotBeUpper`
- `BeLower`
- `NotBeLower`

Fixes #1356

jnyrup commented:
Continuing the discussion about API naming...
The BCL also has `ToTitleCase`, which _does_ use the "Case" suffix, so I'd like to keep the door open for alternative names a bit.
I think `BeUpperCase` reads closer to regular prose.
While we should get inspiration from BCL, it does not need to dictate our choices.

How does `char.IsUpper` handle culture?

I recall that `char.IsUpper` only returns true for characters that are both letters _and_ uppercase.
That clearly justifies that we should have all four combinations, as a character not being upper case does not imply that it is lower case, e.g. digits are neither lower case nor upper case.

I think the test suite would benefit from tests like these to reflect those cases.
```cs
subject = "A1";

Action act = () => subject.Should().BeUpper();

act.Should().Throw();
```

```cs
subject = "a1";

Action act = () => subject.Should().NotBeUpper();

act.Should().NotThrow();
```

and the complementary for `[Not]BeLower`.

dennisdoomen commented on Src/FluentAssertions/Primitives/StringAssertions.cs:
```diff
@@ -1114,6 +1114,86 @@ public AndConstraint<TAssertions> BeNullOrWhiteSpace(string because = "", params
         }
 
+        /// <summary>
+        /// Asserts that a string is uppercase.
```
I think it would be better to phrase it like "Asserts that all characters in a string are in upper casing"

dennisdoomen commented on Src/FluentAssertions/Primitives/StringAssertions.cs:
```diff
@@ -1114,6 +1114,86 @@ public AndConstraint<TAssertions> BeNullOrWhiteSpace(string because = "", params
+        /// <param name="becauseArgs">
+        /// Zero or more objects to format using the placeholders in <paramref name="because" />.
+        /// </param>
+        public AndConstraint<TAssertions> BeUpper(string because = "", params object[] becauseArgs)
```
Maybe `BeUpperCased`?

dennisdoomen commented on Tests/FluentAssertions.Specs/Primitives/StringAssertionSpecs.cs:
```diff
@@ -3243,5 +3243,220 @@ public void When_a_whitespace_string_is_expected_to_not_be_null_or_whitespace_it
+        #region (Not) Upper
+
+        [Fact]
+        public void Should_succeed_when_asserting_upper_string_to_be_upper()
```
Our convention is `When_something_it_should_do_something`. I know we have some older tests that don't follow that convention, but we need to change that,

weitzhandler replied:
Do you mean `When_asserting_uppercase_string_to_be_upper_should_succeed`, in this case? Just to be clear.

dennisdoomen replied:
Almost ;-)

`When_an_upper_case_string_is_supposed_to_be_in_upper_case_only_it_should_not_throw`

dennisdoomen commented on Src/FluentAssertions/Primitives/StringAssertions.cs:
```diff
@@ -1114,6 +1114,86 @@ public AndConstraint<TAssertions> BeNullOrWhiteSpace(string because = "", params
+        public AndConstraint<TAssertions> BeUpper(string because = "", params object[] becauseArgs)
+        {
+            Execute.Assertion
+                .ForCondition(Subject?.All(char.IsUpper) == true)
```
What if the string is empty?

weitzhandler replied:
That means it's not upper case. And I've actually asked this question [here](https://github.com/fluentassertions/fluentassertions/issues/1356#issuecomment-659963900).

dennisdoomen commented on Src/FluentAssertions/Primitives/StringAssertions.cs:
```diff
@@ -1114,6 +1114,86 @@ public AndConstraint<TAssertions> BeNullOrWhiteSpace(string because = "", params
+            Execute.Assertion
+                .ForCondition(Subject?.All(char.IsUpper) == true)
+                .BecauseOf(because, becauseArgs)
+                .FailWith("Expected {context:string} to be upper{reason}, but found {0}.", Subject);
```
I would be more specific, e.g. `Expected all characters in {context:string} to be upper cased`

weitzhandler commented:
https://github.com/fluentassertions/fluentassertions/commit/cca1d01bd64121dca56783c2ff4dbabc25077d59:

- Resolved comments
- Rebased on `upstream/develop`
- Updated docs

jnyrup commented on Src/FluentAssertions/Primitives/StringAssertions.cs:
```diff
@@ -1114,6 +1114,86 @@ public AndConstraint<TAssertions> BeNullOrWhiteSpace(string because = "", params
+        }
+
+        /// <summary>
+        /// Asserts that all characters in a string are not in upper casing.
```
```suggestion
        /// Asserts that all characters in a string are not in lower casing.
```

jnyrup commented on docs/_pages/releases.md:
```diff
@@ -34,6 +34,7 @@ sidebar:
 * Added `AllBe` to `StringCollectionAssertions` to be able to assert that all strings in collection are equal to the specified string - [#1332](https://github.com/fluentassertions/fluentassertions/pull/1332).
 * Added `ForConstraint` method to `AssertionScope` to open up `OccurenceConstraint` for usage in custom assertion extensions - [#1341](https://github.com/fluentassertions/fluentassertions/pull/1341).
 * Added `NotContainInOrder` to `CollectionAssertions` and `StringCollectionAssertions` to be able to assert that the collection does not contain the specified elements in the exact same order, not necessarily consecutive - [#1339](https://github.com/fluentassertions/fluentassertions/pull/1339).
+* Added `[Not]BeUpperCase` and `[Not]BeLowerCase` to `StringAssertions` to be able to assert that a string is in upper or lower casing or not - [#1357](https://github.com/fluentassertions/fluentassertions/pull/1357).
```
```suggestion
* Added `[Not]BeUpperCased` and `[Not]BeLowerCased` to `StringAssertions` to be able to assert that a string is in upper or lower casing or not - [#1357](https://github.com/fluentassertions/fluentassertions/pull/1357).
```

jnyrup commented on docs/_pages/strings.md:
```diff
@@ -18,6 +18,10 @@ theString.Should().NotBeEmpty("because the string is not empty");
 theString.Should().HaveLength(0);
 theString.Should().BeNullOrWhiteSpace(); // either null, empty or whitespace only
 theString.Should().NotBeNullOrWhiteSpace();
+theString.Should().BeUpperCase();
+theString.Should().NotBeUpperCase();
+theString.Should().BeLowerCase();
+theString.Should().NotBeLowerCase();
```
```suggestion
theString.Should().BeUpperCased();
theString.Should().NotBeUpperCased();
theString.Should().BeLowerCased();
theString.Should().NotBeLowerCased();
```
